main/src/ca/uwaterloo/flix/language/phase/PatternExhaustiveness.scala

```diff
@@ -0,0 +1,524 @@
+       * root constructor set, or if we match every constructor for the
+       * expression
+       */
+      expCtors.foldRight(List.empty[TypeConstructor])((x, xs) => if (ctors.exists(y => sameCtor(x, y))) xs else x :: xs)
+    }
+
+    /**
+      * Gets the number of arguments for a constructor, takes into account the "fake constructors"
+      *
+      * @param ctor The constructor to get from
+      * @return The number of arguments for the constructor
+      */
+    def countCtorArgs(ctor: TypeConstructor): Int = ctor match {
+      case True() => 0
+      case False() => 0
+      case Unit() => 0
+      case Tuple(args) => args.size
+      case BigInt() => 0
+      case Int8() => 0
+      case Int16() => 0
+      case Int32() => 0
+      case Int64() => 0
+      case Float32() => 0
+      case Float64() => 0
+      case Str() => 0
+      case Char() => 0
+      case Wild() => 0
+      case Enum(_, _, numArgs, _) => numArgs
+    }
+
+    def countTypeArgs(tpe: Type): Int = tpe match {
+      case Type.Var(id, kind) => 0
+      case Type.Unit => 0
+      case Bool => 0
+      case Type.Char => 0
+      case Type.Float32 => 0
+      case Type.Float64 => 0
+      case Type.Int8 => 0
+      case Type.Int16 => 0
+      case Type.Int32 => 0
+      case Type.Int64 => 0
+      case Type.BigInt => 0
+      case Type.Str => 0
+      case Native => 0
+      case Ref => 0
+      case Arrow(length) => length
+      case FTuple(length) => length
+      case Type.Enum(sym, kind) => 0
+      case Apply(t, ts) => countTypeArgs(t)
+    }
+
+    def prettyPrintCtor(ctor: TypeConstructor): String = ctor match {
+      case True() => "True"
+      case False() => "False"
+      case Unit() => "Unit"
+      case Tuple(args) => "(" + args.foldRight("")((x, xs) => if (xs == "") prettyPrintCtor(x) + xs else prettyPrintCtor(x) + ", " + xs) + ")"
+      case BigInt() => "BigInt"
+      case Int8() => "Int8"
+      case Int16() => "Int16"
+      case Int32() => "Int32"
+      case Int64() => "Int64"
+      case Float32() => "Float32"
+      case Float64() => "Float64"
+      case Str() => "Str"
+      case Char() => "Char"
+      case Wild() => "_"
+      case Enum(name, _, num_args, args) => if (num_args == 0 ) name else name + prettyPrintCtor(Tuple(args))
+    }
+
+
+    /**
+      * Checks if two TypeConstructors refers to the same constructor.
+      *
+      * @param c1 First constructor to compare
+      * @param c2 Second constructor to compare
+      * @return True if they are the same constructor
+      */
+    def sameCtor(c1: TypeConstructor, c2: TypeConstructor): Boolean = (c1,c2) match {
+      // Two enums are the same constructor if they have the same name and enum sym
+      case (TypeConstructor.Enum(n1, s1, _, _), TypeConstructor.Enum(n2, s2, _, _)) => n1 == n2 && s1 == s2
+      // Everything else is the same constructor if they are the same type
+      case (a, b) =>  a.getClass == b.getClass;
```

Comment:
You should definitely never be using `getClass` like this. What test are you trying to do?